go/libraries/doltcore/branch_control/branch_control.go

```diff
@@ -228,28 +256,58 @@ func CanCreateBranch(ctx context.Context, branchName string) error {
 // However, not all CLI commands use *sql.Context, and therefore will not have any user associated with the context. In
 // these cases, CanDeleteBranch will pass as we want to allow all local commands to freely delete branches.
 func CanDeleteBranch(ctx context.Context, branchName string) error {
-	if !enabled {
-		return nil
-	}
 	branchAwareSession := GetBranchAwareSession(ctx)
 	// A nil session means we're not in the SQL context, so we allow the delete operation
 	if branchAwareSession == nil {
 		return nil
 	}
-	StaticController.Access.RWMutex.RLock()
-	defer StaticController.Access.RWMutex.RUnlock()
+	controller := branchAwareSession.GetController()
+	// Any context that has a non-nil session should always have a non-nil controller, so this is an error
+	if controller == nil {
+		return ErrMissingController.New()
+	}
+	controller.Access.RWMutex.RLock()
+	defer controller.Access.RWMutex.RUnlock()
 
 	user := branchAwareSession.GetUser()
 	host := branchAwareSession.GetHost()
 	// Get the permissions for the branch, user, and host combination
-	_, perms := StaticController.Access.Match(branchName, user, host)
+	_, perms := controller.Access.Match(branchName, user, host)
 	// If the user has the write or admin flags, then we allow access
 	if (perms&Permissions_Write == Permissions_Write) || (perms&Permissions_Admin == Permissions_Admin) {
 		return nil
 	}
 	return ErrCannotDeleteBranch.New(user, host, branchName)
 }
 
+// AddAccessEntryForContext adds an entry in the access table for the user represented by the given context. If the
+// context is missing some functionality that is needed to perform the addition, such as a user or the Controller, then
+// this simply returns.
+func AddAccessEntryForContext(ctx context.Context, branchName string) error {
+	branchAwareSession := GetBranchAwareSession(ctx)
+	if branchAwareSession == nil {
+		return nil
+	}
+	controller := branchAwareSession.GetController()
+	if controller == nil {
+		return nil
+	}
+
+	user := branchAwareSession.GetUser()
+	host := branchAwareSession.GetHost()
+	// Check if we already have admin permissions for the given branch, as there's no need to do another insertion if so
+	controller.Access.RWMutex.RLock()
```

Comment:
This is a bad pattern. The access object should manage its own concurrency needs without requiring locking on the outside like this.

Reply:
I can take another pass at it, but it definitely complicated the model. For example, when inserting a row, we also read the table. Right now the insertion function on the table just gets a `Lock`. Changing it so that each function acquires a lock means you can get into a state where we read a row to update and it's changed by another user before we can do anything with it. So the way around that is to basically move all of the dolt_table logic to the `Access` struct, and same for `Namespace`.

Like if no `Access` functions acquire a lock (except `Serialize/Deserialize` which won't be nested in anything), then we can easily manage it from the layers above (`branch_control.go` and the dolt_tables). I really think this keeps everything a bit more straightforward.